servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherBuffer.java

```diff
@@ -190,57 +222,113 @@ public void onComplete() {
                 // external entity. This means we neither re-enter this method nor the items source can terminate
                 // when we are inside this method (onNext and onError/onComplete can not be concurrent).
-                if (cMaybeAccumulator == TERMINATED || cMaybeAccumulator instanceof ItemsTerminated) {
+                if (cMaybeAccumulator == TERMINATED) {
                     return;
                 }
-                assert cMaybeAccumulator != ADDING;
+                assert cMaybeAccumulator != ADDING; // Invocation of 'accumulate' method is expected to be sequential
                 // If we are ADDING and maybeAccumulator has changed then either it should have terminated or
-                // a new accumulator has been received. If a new accumulator is received, we could have never added to
-                // that accumulator (as we were adding to an old accumulator), so we discard that accumulator.
+                // a new accumulator has been received. If a new accumulator is received, we will finish the current
+                // accumulator and emit the result.
                 if (maybeAccumulatorUpdater.compareAndSet(this, cMaybeAccumulator, ADDING)) {
+                    // Use the next accumulator if the current one is already on the way to the target:
                     @SuppressWarnings("unchecked")
-                    final Accumulator<T, B> accumulator = (Accumulator<T, B>) cMaybeAccumulator;
+                    final Accumulator<T, B> accumulator =
+                            NextAccumulatorHolder.class.equals(cMaybeAccumulator.getClass()) ?
+                                    ((NextAccumulatorHolder<T, B>) cMaybeAccumulator).accumulator :
+                                    (Accumulator<T, B>) cMaybeAccumulator;
                     accumulator.accumulate(item);
-                    // If maybeAccumulator changed, it means either we terminated or we receive a new accumulator.
-                    // We could not have added any new items to the newly received accumulator, so we overwrite the
-                    // accumulator, effectively discarding the new accumulator. We will emit this accumulator on the
-                    // next boundary emission.
-                    maybeAccumulatorUpdater.accumulateAndGet(this, accumulator,
-                            (prev, next) -> prev == TERMINATED ? TERMINATED : next);
+                    final Object nextState = maybeAccumulatorUpdater.accumulateAndGet(this, accumulator,
+                            (prev, next) -> prev == ADDING ? next : prev);
+                    if (nextState == accumulator || nextState == TERMINATED) {
+                        return;
+                    }
+                    // Received the next boundary while adding, deliver current accumulator and unwrap the next boundary
+                    assert nextState instanceof NextAccumulatorHolder : "Expected next boundary";
```

Comment:
nit: we will get a class cast on the next line anyways, consider removing this assertion.